content/blog/2020/08/2020-08-29-git-performance-improvement-phase3.adoc

```diff
@@ -0,0 +1,126 @@
+---
+layout: post
+title: "Git Plugin Performance Improvement: Final Phase and Release"
+tags:
+- plugins
+- git
+- platform-sig
+- developer
+- performance
+- gsoc
+- gsoc2020
+author: rishabhbudhouliya
+opengraph:
+  image: /images/post-images/2020/07-git-plugin-performance-improvement/gsoc-git-opengraph.png
+---
+
+Since the beginning of the project, the core value which drove its progress was "To enhance the user experience for running Jenkins jobs by reducing the overall execution time".
+
+To achieve this goal, we laid out a path: 
+
+* Compare the two existing git implementations i.e CliGitAPIImpl and JGitAPIImpl using performance benchmarking
+* Use the results to create a feature which would improve the overall performance of git plugin
+* Also, fix existing user reported performance issues
+
+Let's take a journey to understand how we've build the new features. If you'd like to skip the journey part, you can directly skip to the end section to see what we've done!
+
+== Journey to the release of Git Plugin 4.4.0 and Git Client Plugin 3.4.0
+
+The project started with deciding to choose a git operation and then trying to compare the performance of that operation by using `command line git` and then with `JGit`.
+
+=== Stage 1: Benchmark results with git fetch
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-fetch-results.png[git-fetch-results, width=938]
+
+* The performance of git fetch (average execution time/op) is strongly correlated to the size of a repository
+* There exists an inflection point on the scale of repository size after which the nature of `JGit` performance changes (it starts to degrade)
+* After running multiple benchmarks, it is safe to say that for a large sized repository `command line git` would be a better choice of implementation.
+* We can use this insight to implement a feature which avoids `JGit` when it comes to large repositories.
+
+=== Stage 2: Validate the benchmark results across multiple platforms
+
+Running git fetch operation for a 400 MiB sized repository on:
```

Comment:
Needs an introductory paragraph to describe why the project was concerned about platform dependencies.
```suggestion
The project was also concerned that there might be important differences between operating systems.
For example, what if command line Git for Windows performed very differently than command line Git on Linux or FreeBSD?

Benchmarks were run to compare fetch performance on several platforms.
Running git fetch operation for a 400 MiB sized repository on:
```